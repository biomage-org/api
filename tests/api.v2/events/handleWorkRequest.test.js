const handleWorkRequest = require('../../../src/api.v2/events/handleWorkRequest');
const fake = require('../../test-utils/constants');

jest.mock('../../../src/api.v2/middlewares/authMiddlewares');
jest.mock('aws-xray-sdk');
jest.mock('../../../src/api.v2/events/validateAndSubmitWork', () => (data) => {
  if (data.uuid === 'invalid') {
    throw new Error('invalid request');
  }
  return jest.fn();
});


const data = {
  uuid: 'someuuid-asd-asd-asdddasa',
  Authorization: 'Bearer some-token',
  experimentId: fake.EXPERIMENT_ID,
};

describe('Handle work socket callback', () => {
  const mockSocket = { emit: jest.fn() };
  const mockXraySegment = { addError: jest.fn() };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('submits work', async () => {
    await handleWorkRequest(mockSocket, data, mockXraySegment);
    expect(mockSocket.emit.mock.calls[0]).toMatchSnapshot();
  });

  it('Failing submits a work-response with error', async () => {
    data.uuid = 'invalid';
    await handleWorkRequest(mockSocket, data, mockXraySegment);
    expect(mockSocket.emit.mock.calls[0]).toMatchSnapshot();
    expect(mockXraySegment.addError).toHaveBeenCalled();
  });
});
