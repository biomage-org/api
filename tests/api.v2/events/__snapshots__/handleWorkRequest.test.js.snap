// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle work socket callback Failing submits a work-response with error 1`] = `
Array [
  "WorkResponse-invalid",
  Object {
    "request": Object {
      "Authorization": "Bearer some-token",
      "experimentId": "experimentid11111111111111111111",
      "uuid": "invalid",
    },
    "response": Object {
      "cacheable": false,
      "error": "invalid request",
      "trace": undefined,
    },
    "results": Array [],
  },
]
`;

exports[`Handle work socket callback data without authorization also throws an error 1`] = `
Array [
  "WorkResponse-invalid",
  Object {
    "request": Object {
      "experimentId": "experimentid11111111111111111111",
      "uuid": "invalid",
    },
    "response": Object {
      "cacheable": false,
      "error": "Authentication token must be present.",
      "trace": undefined,
    },
    "results": Array [],
  },
]
`;

exports[`Handle work socket callback submits work 1`] = `
Array [
  "WorkerInfo-experimentid11111111111111111111",
  Object {
    "response": Object {
      "podInfo": [MockFunction],
      "trace": undefined,
    },
  },
]
`;
